File: backend/src/main/java/cl/duoc/sffe/service/ViajeService.java
package cl.duoc.sffe.service;

import cl.duoc.sffe.dto.MascotaRequest;
import cl.duoc.sffe.dto.MenorRequest;
import cl.duoc.sffe.dto.SagRequest;
import cl.duoc.sffe.dto.VehiculoRequest;
import cl.duoc.sffe.dto.ViajeRequest;
import cl.duoc.sffe.dto.ViajeResponse;
import cl.duoc.sffe.exception.AuthException;
import cl.duoc.sffe.exception.ViajeException;
import cl.duoc.sffe.model.DeclaracionSag;
import cl.duoc.sffe.model.Mascota;
import cl.duoc.sffe.model.Menor;
import cl.duoc.sffe.model.Usuario;
import cl.duoc.sffe.model.Vehiculo;
import cl.duoc.sffe.model.Viaje;
import cl.duoc.sffe.repository.DeclaracionSagRepository;
import cl.duoc.sffe.repository.MascotaRepository;
import cl.duoc.sffe.repository.MenorRepository;
import cl.duoc.sffe.repository.UsuarioRepository;
import cl.duoc.sffe.repository.VehiculoRepository;
import cl.duoc.sffe.repository.ViajeRepository;
import cl.duoc.sffe.util.DocumentoValidator;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDate;
import java.util.List;

/**
 * Lógica de negocio de expedientes de viaje: itinerario, menores, vehículo y
 * declaración SAG (RF02, RF03, RF04). Todas las operaciones se acotan al
 * usuario autenticado, identificado por su identificador a partir del JWT.
 */
@Service
public class ViajeService {

    private final ViajeRepository viajeRepository;
    private final MenorRepository menorRepository;
    private final VehiculoRepository vehiculoRepository;
    private final MascotaRepository mascotaRepository;
    private final DeclaracionSagRepository declaracionSagRepository;
    private final UsuarioRepository usuarioRepository;
    private final DocumentoValidator documentoValidator;
    private final FileStorageService fileStorageService;

    public ViajeService(ViajeRepository viajeRepository,
                        MenorRepository menorRepository,
                        VehiculoRepository vehiculoRepository,
                        MascotaRepository mascotaRepository,
                        DeclaracionSagRepository declaracionSagRepository,
                        UsuarioRepository usuarioRepository,
                        DocumentoValidator documentoValidator,
                        FileStorageService fileStorageService) {
        this.viajeRepository = viajeRepository;
        this.menorRepository = menorRepository;
        this.vehiculoRepository = vehiculoRepository;
        this.mascotaRepository = mascotaRepository;
        this.declaracionSagRepository = declaracionSagRepository;
        this.usuarioRepository = usuarioRepository;
        this.documentoValidator = documentoValidator;
        this.fileStorageService = fileStorageService;
    }

    /** Crea un nuevo expediente de viaje asociado al usuario autenticado (RF02). */
    @Transactional
    public ViajeResponse crear(String identificador, ViajeRequest request) {
        Usuario usuario = obtenerUsuario(identificador);

        Viaje viaje = Viaje.builder()
                .usuario(usuario)
                .fechaIngreso(request.fechaIngreso())
                .destino(request.destino())
                .paisOrigen(request.paisOrigen())
                .pasoFronterizo(request.pasoFronterizo())
                .motivoViaje(request.motivoViaje())
                .build();

        return ViajeResponse.from(viajeRepository.save(viaje));
    }

    /** Lista los expedientes de viaje del usuario autenticado (RF04). */
    @Transactional(readOnly = true)
    public List<ViajeResponse> misViajes(String identificador) {
        Usuario usuario = obtenerUsuario(identificador);
        return viajeRepository.findByUsuarioIdUsuario(usuario.getIdUsuario())
                .stream()
                .map(ViajeResponse::from)
                .toList();
    }

    /** Obtiene el detalle de un expediente, solo si pertenece al usuario autenticado (RF04). */
    @Transactional(readOnly = true)
    public ViajeResponse obtener(String identificador, Integer idViaje) {
        return ViajeResponse.from(obtenerViajeDelUsuario(identificador, idViaje));
    }

    /** Actualiza los datos del itinerario de un expediente propio (RF02). */
    @Transactional
    public ViajeResponse actualizar(String identificador, Integer idViaje, ViajeRequest request) {
        Viaje viaje = obtenerViajeDelUsuario(identificador, idViaje);

        viaje.setFechaIngreso(request.fechaIngreso());
        viaje.setDestino(request.destino());
        viaje.setPaisOrigen(request.paisOrigen());
        viaje.setPasoFronterizo(request.pasoFronterizo());
        viaje.setMotivoViaje(request.motivoViaje());

        return ViajeResponse.from(viajeRepository.save(viaje));
    }

    /**
     * Agrega un menor de edad al expediente de viaje (RF02). El RUT se valida
     * con el mismo algoritmo módulo 11 que el identificador del usuario, la
     * fecha de nacimiento debe corresponder a una persona menor de 18 años a
     * la fecha de ingreso, y el carnet de identidad y los papeles de
     * antecedentes son obligatorios; el permiso notarial es obligatorio solo
     * cuando {@code requiereAutorizacion = true}.
     */
    @Transactional
    public ViajeResponse agregarMenor(String identificador, Integer idViaje, MenorRequest request,
                                       MultipartFile carnetIdentidad,
                                       MultipartFile papelesAntecedentes,
                                       MultipartFile permisoNotarial) {
        Viaje viaje = obtenerViajeDelUsuario(identificador, idViaje);

        String rutNormalizado = documentoValidator.normalizar(request.rut());
        if (!documentoValidator.validarRut(rutNormalizado)) {
            throw new ViajeException(HttpStatus.BAD_REQUEST, "El RUT del menor no es válido");
        }

        LocalDate fechaNacimiento = request.fechaNacimiento();
        if (fechaNacimiento.isAfter(LocalDate.now())) {
            throw new ViajeException(HttpStatus.BAD_REQUEST,
                    "La fecha de nacimiento del menor no es válida");
        }
        boolean esMenorDeEdad = fechaNacimiento.isAfter(viaje.getFechaIngreso().minusYears(18));
        if (!esMenorDeEdad) {
            throw new ViajeException(HttpStatus.BAD_REQUEST,
                    "La fecha de nacimiento indicada corresponde a una persona mayor de edad");
        }

        boolean requiereAutorizacion =
                request.requiereAutorizacion() != null && request.requiereAutorizacion();

        String carnetPath = fileStorageService.guardarObligatorio(
                carnetIdentidad, "menores", "el carnet de identidad del menor");
        String antecedentesPath = fileStorageService.guardarObligatorio(
                papelesAntecedentes, "menores", "los papeles de antecedentes del menor");
        String permisoPath = requiereAutorizacion
                ? fileStorageService.guardarObligatorio(
                        permisoNotarial, "menores", "el permiso notarial del menor")
                : fileStorageService.guardarOpcional(
                        permisoNotarial, "menores", "el permiso notarial del menor");

        Menor menor = Menor.builder()
                .viaje(viaje)
                .nombre(request.nombre())
                .rut(rutNormalizado)
                .fechaNacimiento(fechaNacimiento)
                .requiereAutorizacion(requiereAutorizacion)
                .carnetIdentidadPath(carnetPath)
                .papelesAntecedentesPath(antecedentesPath)
                .permisoNotarialPath(permisoPath)
                .build();

        menorRepository.save(menor);

        // No se agrega "menor" manualmente a viaje.getMenores(): al ser una
        // colección perezosa, acceder a ella aquí dispara su carga desde BD
        // (ya incluye el menor recién guardado) y duplicaría la entrada.
        return ViajeResponse.from(viaje);
    }

    /**
     * Registra o actualiza un vehículo del expediente (RF03). Un viaje admite
     * como máximo un vehículo principal y un remolque; la operación es un upsert
     * por tipo (busca por {@code esRemolque}, crea si no existe, actualiza si ya
     * existe). El remolque exige que el vehículo principal ya esté registrado.
     * El permiso de circulación es obligatorio para cualquiera de los dos
     * (RF03) y queda visible para Aduana y PDI en el expediente consolidado.
     */
    @Transactional
    public ViajeResponse registrarVehiculo(String identificador, Integer idViaje, VehiculoRequest request,
                                            MultipartFile permisoCirculacion) {
        Viaje viaje = obtenerViajeDelUsuario(identificador, idViaje);

        boolean esRemolque = Boolean.TRUE.equals(request.esRemolque());

        Vehiculo principal = vehiculoRepository
                .findByViajeIdViajeAndEsRemolque(idViaje, false)
                .orElse(null);

        if (esRemolque && principal == null) {
            throw new ViajeException(HttpStatus.CONFLICT,
                    "Debes registrar el vehículo principal antes del remolque");
        }

        if (!esRemolque && (isBlank(request.marca()) || isBlank(request.modelo()) || request.anio() == null)) {
            throw new ViajeException(HttpStatus.BAD_REQUEST,
                    "Marca, modelo y año son obligatorios para el vehículo principal");
        }

        String permisoPath = fileStorageService.guardarObligatorio(
                permisoCirculacion, "vehiculos", "el permiso de circulación del vehículo");

        Vehiculo vehiculo = vehiculoRepository
                .findByViajeIdViajeAndEsRemolque(idViaje, esRemolque)
                .orElseGet(() -> Vehiculo.builder().viaje(viaje).esRemolque(esRemolque).build());

        vehiculo.setPatente(request.patente().toUpperCase());
        vehiculo.setMarca(request.marca());
        vehiculo.setModelo(request.modelo());
        vehiculo.setAnio(request.anio());
        vehiculo.setEsRemolque(esRemolque);
        vehiculo.setVehiculoPrincipalId(esRemolque ? principal.getIdVehiculo() : null);
        vehiculo.setPermisoCirculacionPath(permisoPath);

        vehiculoRepository.save(vehiculo);

        // No se agrega manualmente a viaje.getVehiculos(): la colección perezosa
        // aún no se ha cargado, así que ViajeResponse.from la leerá fresca desde
        // BD (incluyendo el vehículo recién guardado), igual que con los menores.
        return ViajeResponse.from(viaje);
    }

    /**
     * Agrega una mascota al expediente (RF02). El tipo de animal y el número
     * de chip son obligatorios (@NotBlank en {@link MascotaRequest}); el
     * certificado del chip y el carnet de vacunación son obligatorios como
     * archivos adjuntos. Visible para toda fiscalización (Aduana, PDI, SAG).
     */
    @Transactional
    public ViajeResponse agregarMascota(String identificador, Integer idViaje, MascotaRequest request,
                                         MultipartFile certificadoChip, MultipartFile carnetVacunacion) {
        Viaje viaje = obtenerViajeDelUsuario(identificador, idViaje);

        String certificadoPath = fileStorageService.guardarObligatorio(
                certificadoChip, "mascotas", "el certificado del chip de la mascota");
        String vacunacionPath = fileStorageService.guardarObligatorio(
                carnetVacunacion, "mascotas", "el carnet de vacunación de la mascota");

        Mascota mascota = Mascota.builder()
                .viaje(viaje)
                .tipoAnimal(request.tipoAnimal())
                .numeroChip(request.numeroChip())
                .certificadoChipPath(certificadoPath)
                .carnetVacunacionPath(vacunacionPath)
                .build();

        mascotaRepository.save(mascota);

        // No se agrega manualmente a viaje.getMascotas(): igual que con menores
        // y vehículos, ViajeResponse.from relee la colección perezosa desde BD.
        return ViajeResponse.from(viaje);
    }

    /** Guarda o actualiza la Declaración Jurada SAG del expediente, relación 1:1 (RF02). */
    @Transactional
    public ViajeResponse guardarSag(String identificador, Integer idViaje, SagRequest request) {
        Viaje viaje = obtenerViajeDelUsuario(identificador, idViaje);

        boolean declaraDivisas = Boolean.TRUE.equals(request.declaraDivisas());
        if (declaraDivisas
                && (request.montoDivisas() == null
                        || request.montoDivisas().compareTo(java.math.BigDecimal.ZERO) <= 0)) {
            throw new ViajeException(HttpStatus.BAD_REQUEST,
                    "Si declara divisas, debe indicar un monto mayor a cero");
        }

        DeclaracionSag sag = declaracionSagRepository.findByViajeIdViaje(idViaje)
                .orElseGet(() -> DeclaracionSag.builder().viaje(viaje).build());

        // Sección SAG
        sag.setDeclaraProductos(request.declaraProductos());
        sag.setProductos(request.productos());

        // Sección Aduanas: si no declara, se descartan los detalles asociados.
        sag.setDeclaraDivisas(declaraDivisas);
        sag.setMontoDivisas(declaraDivisas ? request.montoDivisas() : null);
        sag.setMonedaDivisas(declaraDivisas ? request.monedaDivisas() : null);

        boolean declaraMercancias = Boolean.TRUE.equals(request.declaraMercancias());
        sag.setDeclaraMercancias(declaraMercancias);
        sag.setDetalleMercancias(declaraMercancias ? request.detalleMercancias() : null);

        viaje.setDeclaracionSag(declaracionSagRepository.save(sag));

        return ViajeResponse.from(viaje);
    }

    private boolean isBlank(String valor) {
        return valor == null || valor.isBlank();
    }

    private Usuario obtenerUsuario(String identificador) {
        return usuarioRepository.findByIdentificador(identificador)
                .orElseThrow(() -> new AuthException(
                        HttpStatus.UNAUTHORIZED, "El usuario de la sesión no existe"));
    }

    /** Busca el viaje por id y valida que pertenezca al usuario autenticado. */
    private Viaje obtenerViajeDelUsuario(String identificador, Integer idViaje) {
        Usuario usuario = obtenerUsuario(identificador);

        Viaje viaje = viajeRepository.findById(idViaje)
                .orElseThrow(() -> new ViajeException(
                        HttpStatus.NOT_FOUND, "El expediente de viaje no existe"));

        if (!viaje.getUsuario().getIdUsuario().equals(usuario.getIdUsuario())) {
            throw new ViajeException(
                    HttpStatus.FORBIDDEN, "No tienes acceso a este expediente de viaje");
        }

        return viaje;
    }
}

File: frontend/src/services/viajeService.ts
import api from './api'

export type EstadoViaje = 'PENDIENTE' | 'APROBADO' | 'RECHAZADO'
export type EstadoDeclaracion = 'PENDIENTE' | 'VALIDADO' | 'RECHAZADO'

export interface ViajePayload {
  fechaIngreso: string
  destino: string
  pasoFronterizo: string
  paisOrigen?: string | null
  motivoViaje: string
}

export interface MenorPayload {
  nombre: string
  rut: string
  fechaNacimiento: string
  requiereAutorizacion: boolean
}

export interface MenorInfo extends MenorPayload {
  idMenor: number
}

/**
 * Archivos adjuntos del menor (RF02). El carnet y los papeles de
 * antecedentes son obligatorios; el permiso notarial es obligatorio solo
 * cuando {@code requiereAutorizacion = true} (validado en el backend).
 */
export interface MenorArchivos {
  carnetIdentidad: File | null
  papelesAntecedentes: File | null
  permisoNotarial: File | null
}

/** Datos de una mascota del viaje (RF02). */
export interface MascotaPayload {
  tipoAnimal: string
  numeroChip: string
}

export interface MascotaInfo extends MascotaPayload {
  idMascota: number
  certificadoChip: boolean
  carnetVacunacion: boolean
}

/** Certificado del chip y carnet de vacunación de la mascota (RF02): ambos obligatorios. */
export interface MascotaArchivos {
  certificadoChip: File | null
  carnetVacunacion: File | null
}

export type EstadoQr = 'ACTIVO' | 'USADO' | 'EXPIRADO'

export interface VehiculoPayload {
  patente: string
  marca?: string | null
  modelo?: string | null
  anio?: number | null
  /** true si es carro de arrastre o remolque (CAMBIO 6.1). */
  esRemolque?: boolean
}

export interface VehiculoInfo extends VehiculoPayload {
  idVehiculo: number
  esRemolque: boolean
  /** true si el permiso de circulación fue adjuntado (RF03, obligatorio). */
  permisoCirculacion: boolean
}

export interface SagPayload {
  // Sección SAG
  declaraProductos: boolean
  productos: string
  // Sección Aduanas
  declaraDivisas: boolean
  montoDivisas?: number | null
  monedaDivisas?: string | null
  declaraMercancias: boolean
  detalleMercancias?: string | null
}

export interface SagInfo extends SagPayload {
  idDeclaracion: number
  estado: EstadoDeclaracion
  firmaDigital: string | null
  fecha: string
}

/** Código QR anidado en el expediente (RF04). */
export interface QrInfo {
  codigo: string
  estado: EstadoQr
  fechaGeneracion: string
}

export interface Viaje extends ViajePayload {
  idViaje: number
  estado: EstadoViaje
  /** Motivo detallado del rechazo (RF05), visible en el ticket del pasajero. Nulo salvo RECHAZADO. */
  motivoRechazo: string | null
  createdAt: string
  /** true si el pasajero adjuntó su carnet al registrarse; las cuentas semilla no lo tienen. */
  carnetIdentidad: boolean
  /** true si el pasajero adjuntó sus papeles de antecedentes al registrarse. */
  papelesAntecedentes: boolean
  /** Lista de vehículos del viaje: principal y, opcionalmente, remolque (1:N). */
  vehiculos: VehiculoInfo[]
  /** Mascotas del viaje (RF02), visibles para toda fiscalización. */
  mascotas: MascotaInfo[]
  sag: SagInfo | null
  menores: MenorInfo[]
  qr: QrInfo | null
}

/** Vehículo principal del viaje (no remolque), o null si no hay. */
export function vehiculoPrincipal(viaje: Viaje): VehiculoInfo | null {
  return viaje.vehiculos.find((v) => !v.esRemolque) ?? null
}

/** Remolque / carro de arrastre del viaje, o null si no hay. */
export function vehiculoRemolque(viaje: Viaje): VehiculoInfo | null {
  return viaje.vehiculos.find((v) => v.esRemolque) ?? null
}

/** Formatea el número de expediente como EXP-00042 (CAMBIO 6.1). */
export function numeroExpediente(idViaje: number): string {
  return `EXP-${String(idViaje).padStart(5, '0')}`
}

/** POST /api/viajes — crea un nuevo expediente de viaje (RF02). */
export async function crearViaje(payload: ViajePayload): Promise<Viaje> {
  const { data } = await api.post<Viaje>('/viajes', payload)
  return data
}

/** PUT /api/viajes/{id} — actualiza el itinerario de un expediente propio (RF02). */
export async function actualizarViaje(
  idViaje: number,
  payload: ViajePayload,
): Promise<Viaje> {
  const { data } = await api.put<Viaje>(`/viajes/${idViaje}`, payload)
  return data
}

/** GET /api/viajes/mis-viajes — expedientes del usuario autenticado (RF04). */
export async function misViajes(): Promise<Viaje[]> {
  const { data } = await api.get<Viaje[]>('/viajes/mis-viajes')
  return data
}

/** GET /api/viajes/{id} — detalle de un expediente propio (RF04). */
export async function obtenerViaje(idViaje: number): Promise<Viaje> {
  const { data } = await api.get<Viaje>(`/viajes/${idViaje}`)
  return data
}

/**
 * POST /api/viajes/{id}/menores — agrega un menor de edad al expediente
 * (RF02). Multipart: la parte "datos" lleva el JSON del menor; el carnet de
 * identidad y los papeles de antecedentes son obligatorios, y el permiso
 * notarial lo es solo si requiereAutorizacion es true.
 */
export async function agregarMenor(
  idViaje: number,
  payload: MenorPayload,
  archivos: MenorArchivos,
): Promise<Viaje> {
  const formData = new FormData()
  formData.append('datos', new Blob([JSON.stringify(payload)], { type: 'application/json' }))
  if (archivos.carnetIdentidad) {
    formData.append('carnetIdentidad', archivos.carnetIdentidad)
  }
  if (archivos.papelesAntecedentes) {
    formData.append('papelesAntecedentes', archivos.papelesAntecedentes)
  }
  if (archivos.permisoNotarial) {
    formData.append('permisoNotarial', archivos.permisoNotarial)
  }
  const { data } = await api.post<Viaje>(`/viajes/${idViaje}/menores`, formData)
  return data
}

/**
 * PUT /api/viajes/{id}/menores/{idMenor} — actualiza nombre, RUT, fecha de
 * nacimiento y autorización notarial de un menor ya guardado (RF02), sin
 * tocar sus archivos adjuntos (para eso está reemplazarArchivoMenor).
 */
export async function actualizarMenor(
  idViaje: number,
  idMenor: number,
  payload: MenorPayload,
): Promise<Viaje> {
  const { data } = await api.put<Viaje>(`/viajes/${idViaje}/menores/${idMenor}`, payload)
  return data
}

/** DELETE /api/viajes/{id}/menores/{idMenor} — quita un menor del expediente, definitivamente (RF02). Solo mientras el viaje sigue PENDIENTE. */
export async function eliminarMenor(idViaje: number, idMenor: number): Promise<Viaje> {
  const { data } = await api.delete<Viaje>(`/viajes/${idViaje}/menores/${idMenor}`)
  return data
}

/**
 * POST /api/viajes/{id}/vehiculo — registra o actualiza el vehículo del
 * expediente (RF03). Multipart: la parte "datos" lleva el JSON del vehículo;
 * el permiso de circulación es obligatorio (principal o remolque), visible
 * para Aduana y PDI.
 */
export async function registrarVehiculo(
  idViaje: number,
  payload: VehiculoPayload,
  permisoCirculacion: File | null,
): Promise<Viaje> {
  const formData = new FormData()
  formData.append('datos', new Blob([JSON.stringify(payload)], { type: 'application/json' }))
  if (permisoCirculacion) {
    formData.append('permisoCirculacion', permisoCirculacion)
  }
  const { data } = await api.post<Viaje>(`/viajes/${idViaje}/vehiculo`, formData)
  return data
}

/**
 * POST /api/viajes/{id}/mascotas — agrega una mascota al expediente (RF02).
 * Multipart: la parte "datos" lleva el JSON de la mascota; el certificado
 * del chip y el carnet de vacunación son obligatorios.
 */
export async function agregarMascota(
  idViaje: number,
  payload: MascotaPayload,
  archivos: MascotaArchivos,
): Promise<Viaje> {
  const formData = new FormData()
  formData.append('datos', new Blob([JSON.stringify(payload)], { type: 'application/json' }))
  if (archivos.certificadoChip) {
    formData.append('certificadoChip', archivos.certificadoChip)
  }
  if (archivos.carnetVacunacion) {
    formData.append('carnetVacunacion', archivos.carnetVacunacion)
  }
  const { data } = await api.post<Viaje>(`/viajes/${idViaje}/mascotas`, formData)
  return data
}

/** PUT /api/viajes/{id}/mascotas/{idMascota} — actualiza tipo de animal y número de chip, sin tocar sus archivos (RF02). */
export async function actualizarMascota(
  idViaje: number,
  idMascota: number,
  payload: MascotaPayload,
): Promise<Viaje> {
  const { data } = await api.put<Viaje>(`/viajes/${idViaje}/mascotas/${idMascota}`, payload)
  return data
}

/** DELETE /api/viajes/{id}/mascotas/{idMascota} — quita una mascota del expediente, definitivamente (RF02). Solo mientras el viaje sigue PENDIENTE. */
export async function eliminarMascota(idViaje: number, idMascota: number): Promise<Viaje> {
  const { data } = await api.delete<Viaje>(`/viajes/${idViaje}/mascotas/${idMascota}`)
  return data
}

/** POST /api/viajes/{id}/sag — guarda o actualiza la Declaración Jurada SAG (RF02). */
export async function guardarSag(idViaje: number, payload: SagPayload): Promise<Viaje> {
  const { data } = await api.post<Viaje>(`/viajes/${idViaje}/sag`, payload)
  return data
}

/**
 * PUT /api/viajes/{id}/archivos/... — reemplaza un archivo ya subido (por
 * ejemplo, si se adjuntó el documento equivocado). Solo permitido mientras
 * el viaje sigue PENDIENTE.
 */
async function subirReemplazo(url: string, archivo: File): Promise<void> {
  const formData = new FormData()
  formData.append('archivo', archivo)
  await api.put(url, formData)
}

export function reemplazarArchivoUsuario(idViaje: number, campo: string, archivo: File): Promise<void> {
  return subirReemplazo(`/viajes/${idViaje}/archivos/usuario/${campo}`, archivo)
}

export function reemplazarArchivoMenor(
  idViaje: number, idMenor: number, campo: string, archivo: File,
): Promise<void> {
  return subirReemplazo(`/viajes/${idViaje}/archivos/menores/${idMenor}/${campo}`, archivo)
}

export function reemplazarArchivoVehiculo(
  idViaje: number, idVehiculo: number, campo: string, archivo: File,
): Promise<void> {
  return subirReemplazo(`/viajes/${idViaje}/archivos/vehiculos/${idVehiculo}/${campo}`, archivo)
}

export function reemplazarArchivoMascota(
  idViaje: number, idMascota: number, campo: string, archivo: File,
): Promise<void> {
  return subirReemplazo(`/viajes/${idViaje}/archivos/mascotas/${idMascota}/${campo}`, archivo)
}

const STORAGE_ID_VIAJE = 'sffe_id_viaje_activo'

/** Recupera el id del expediente de viaje en curso (flujo de registro de varios pasos). */
export function getIdViajeActivo(): number | null {
  const valor = localStorage.getItem(STORAGE_ID_VIAJE)
  return valor ? Number(valor) : null
}

/** Guarda el id del expediente de viaje en curso. */
export function setIdViajeActivo(idViaje: number): void {
  localStorage.setItem(STORAGE_ID_VIAJE, String(idViaje))
}
